### DISCUSSION TITLE
Adding the PI Search feature

### DISCUSSION THREAD

**Discussion by Yurij20** [PE5PVB/TEF6686_ESP32#600] [Ideas]
I suggest adding the PI search function.  It should work under the following conditions: if it is turned on through the menu, and when switching saved stations on the switched cell there is no or weak reception and the PI code of the saved station is stored on it, then the search for the same station is activated using this PI code.  This is implemented on car radios as far as you know.

**Comment by PE5PVB** [comment]
Problem is that this can take up to at least 20 seconds. This is not user friendly. To avoid this I should store all AF's for each preset, which is a hugh demand on the memory.

**Reply by Yurij20** [reply]
> Problem is that this can take up to at least 20 seconds. This is not user friendly. To avoid this I should store all AF's for each preset, which is a hugh demand on the memory.

I see) It’s strange why on car radios, no matter how many tests I’ve tested, this works simultaneously.  Even if you are in another city and press a preset with the desired station af the frequency of which is not indicated in this city, is it still quickly found by pi?

**Reply by PE5PVB** [reply]
Most have two tuners.

**Reply by Yurij20** [reply]
> Problem is that this can take up to at least 20 seconds. This is not user friendly. To avoid this I should store all AF's for each preset, which is a hugh demand on the memory.

Okay) then I think you can make the pi search function manual by briefly pressing the bottom key in MEM mode.  I agree to accept the fact that this will take 20 seconds, but in manual mode it will not be annoying).  I think the function process should be as follows: When the function is called during the search process, the sound is turned off and an inscription appears on one of the PI Search lines for the duration of the search.  You can interrupt the process by pressing any key or turning the encoder.  If the search result does not find anything you need, then the original screen display before the search is returned.  As far as I know, at this position of the lower key there is already a DX scanning function, so in the menu you can select whether to briefly press the lower key in MEM mode between the start of DX scanning and PI Search :)

**Reply by numerci** [reply]
> Problem is that this can take up to at least 20 seconds. This is not user friendly. To avoid this I should store all AF's for each preset, which is a hugh demand on the memory.

Storing the AF list will not solve the problem. For example we have radio stations here that list in the AF list only the nearest frequencies not all (they have too many). So, if the user goes even further will not have anything useful in the list. So, maybe just a simple scan with PI matching. And this for all stations on memory.

**Reply by PE5PVB** [reply]
Ok. I'll test a simple script to search for a specific PI and see how long it takes.

**Reply by Yurij20** [reply]
> Ok. I'll test a simple script to search for a specific PI and see how long it takes.

As an option for introducing a manual function launch, a better idea appeared: You can add its launch by briefly pressing the encoder in AUTO mode.  This position is still available, just looked

**Comment by PE5PVB** [comment]
Please test https://github.com/PE5PVB/TEF6686_ESP32/commit/0270321b544b667aa3615c613a18ce4599cdab3c

**Reply by Yurij20** [reply]
> Please test [0270321](https://github.com/PE5PVB/TEF6686_ESP32/commit/0270321b544b667aa3615c613a18ce4599cdab3c)

thank you) I'm waiting for the next release, I'll check it)

**Reply by Yurij20** [reply]
> Please test [0270321](https://github.com/PE5PVB/TEF6686_ESP32/commit/0270321b544b667aa3615c613a18ce4599cdab3c)

I tested it, everything works as it should, it finds it clearly by pi :) What I wanted to improve is that in case of unsuccessful pi scanning, the receiver returns to the frequency saved in the current preset from which the scanning began.  Now in this case the device stops at 108.0.  Also make it possible to cancel scanning, for example, by pressing the bottom button and at the same time the receiver also returned to the frequency with which scanning was going on

**Comment by PE5PVB** [comment]
You can already cancel the scan. Problem of returning to last frequency when no AF found is fixed in https://github.com/PE5PVB/TEF6686_ESP32/commit/298488c0dcb2230fb8be1368f174186f9072dae6

**Reply by Yurij20** [reply]
> You can already cancel the scan. Problem of returning to last frequency when no AF found is fixed in [298488c](https://github.com/PE5PVB/TEF6686_ESP32/commit/298488c0dcb2230fb8be1368f174186f9072dae6)



> You can already cancel the scan. Problem of returning to last frequency when no AF found is fixed in [298488c](https://github.com/PE5PVB/TEF6686_ESP32/commit/298488c0dcb2230fb8be1368f174186f9072dae6)

We are waiting for the next firmware version) I will test it

**Comment by Yurij20** [comment]
tested) now everything works as it should. Thank you
